tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/GraphQL/Net/ASM/GraphQLSecurity.cs

```diff
@@ -0,0 +1,176 @@
+                {
+                    object toDuckValue;
+
+                    // For the version 3 and 4 of GraphQL
+                    if (arg.TryDuckCast<ArgumentProxy>(out var argumentV4))
+                    {
+                        name = argumentV4.NameNode.Name;
+                        toDuckValue = argumentV4.Value;
+                    }
+
+                    // For the version 2 of GraphQL
+                    else if (arg.TryDuckCast<ArgumentProxyV2>(out var argumentV2))
+                    {
+                        name = argumentV2.NamedNode.Name;
+                        toDuckValue = argumentV2.Value;
+                    }
+                    else
+                    {
+                        // Failed to cast as argument
+                        break;
+                    }
+
+                    value = GetValue(toDuckValue);
+
+                    object GetValue(object duckVal)
+                    {
+                        // if Value is VariableReference
+                        if (duckVal.TryDuckCast<GraphQLObjectValueProxy>(out var objFieldValue))
+                        {
+                            var dic = new Dictionary<string, object>();
+                            // return objFieldValue.ObjectFields.Select(GetValue).ToList();
+
+                            foreach (var obj in objFieldValue.ObjectFields)
+                            {
+                                if (obj.TryDuckCast<GraphQLObjectFieldProxy>(out var objField))
+                                {
+                                    dic.Add(objField.NameNode.Name, GetValue(objField.Value));
+                                }
+                            }
+
+                            return dic;
+                        }
+                        else if (duckVal.TryDuckCast<GraphQLValuesProxy>(out var argValues))
+                        {
+                            return argValues.Values.Select(GetValue).ToList();
+                        }
+                        else if (duckVal.TryDuckCast<GraphQLValueProxy>(out var argValue))
+                        {
+                            return GetValue(argValue.Value);
+                        }
+                        else if (duckVal.TryDuckCast<VariableReferenceProxy>(out var variableRef))
+                        {
+                            return GetVariableValue(context, variableRef.Name);
+                        }
+
+                        return duckVal.ToString();
+                    }
+                }
+                else
+                {
+                    // An unknown type of Argument is in the AST
+                    break;
+                }
+
+                resolverArguments.Add(name, value);
+            }
+            catch
+            {
+                // Failed to add the argument to the resolver args list
+            }
+        }
+
+        GraphQLSecurityCommon.RegisterResolverCall(scope, resolverName, resolverArguments);
```

Comment:
I may be misunderstanding, it looks like for HotChocolate we call this method each time we encounter a new node, but here we wait to the end? Why this difference?

Reply:
Yes, because this function RegisterResolver() is called each time we encounter a new node in the ExecuteNodeAsync new autoinstrumentation for graphql-dotnet. In HotChocolate, I am not doing that. I go from the root of the AST and do a depth search on it.